src/policy/ancestor_packages.h

```diff
@@ -0,0 +1,135 @@
+                    if (ancestor_subset.size() == rhs.ancestor_subset.size()) {
+                        // Individual feerate. This is not necessarily fee-optimal, but helps in some situations.
+                        // (a.fee * a.vsize > b.fee * a.vsize) is a shortcut for (a.fee / a.vsize > b.fee / b.vsize)
+                        return *fee * *rhs.vsize  > *rhs.fee * *vsize;
+                    }
+                    return ancestor_subset.size() < rhs.ancestor_subset.size();
+                } else {
+                    return mining_sequence.value() < rhs.mining_sequence.value();
+                }
+            }
+        }
+    };
+    /** Map from each txid to PackageEntry */
+    std::map<uint256, PackageEntry> m_txid_to_entry;
+
+    /** Helper function for recursively constructing ancestor caches in ctor. */
+    void visit(const CTransactionRef&);
+public:
+    /** Constructs ancestor package, sorting the transactions topologically and constructing the
+     * txid_to_tx and ancestor_subsets maps. It is ok if the input txns is not sorted.
+     * Expects:
+     * - No duplicate transactions.
+     * - No conflicts between transactions.
+     */
+    AncestorPackage(const Package& m_txns);
+
+    bool IsAncestorPackage() const { return m_is_ancestor_package; }
+    /** Returns all of the transactions, linearized. */
+    Package Txns() const { return m_txns; }
+
+    /** Returns all of the transactions, without the skipped and dangling ones, linearized. */
+    Package FilteredTxns() const;
+    /** Get the sorted, filtered ancestor subpackage for a tx. Includes the tx. Does not
+     * include skipped ancestors. If this transaction dangles, returns std::nullopt. */
+    std::optional<std::vector<CTransactionRef>> GetAncestorSet(const CTransactionRef& tx);
+    /** Get the total fee and vsize of the ancestor subpackage for a tx. Includes the tx. Does not
+     * include skipped ancestors. If this transaction dangles or fee and vsize are
+     * unavailable, returns std::nullopt. This result is always consistent with GetAncestorSet(). */
+    std::optional<std::pair<CAmount, int64_t>> GetAncestorFeeAndVsize(const CTransactionRef& tx);
+    /** Get the fee and vsize of a tx. Returns std::nullopt if this information is unknown. */
+    std::optional<std::pair<CAmount, int64_t>> GetFeeAndVsize(const CTransactionRef& tx) const;
+    /** From now on, skip this tx from any result in GetAncestorSet(). Does not affect Txns().
+     * Should be called when a transaction is accepted to mempool or already found in it. */
+    void Skip(const CTransactionRef& transaction);
+    /** Skip a transaction and all of its descendants. From now on, if this transaction is present
+     * in the ancestor set, GetAncestorSet() returns std::nullopt for that tx. Does not affect Txns().
+     * Should be called when a transaction is missing inputs. */
+    void SkipWithDescendants(const CTransactionRef& transaction);
```

Comment:
I think you thought about it in term of `AncestorPackage` API design, though what is the purpose of `SkipWithDescendants` as the “dangle” state of a transaction could be determined as soon as all the transactions are received and `IsPackageWellFormed` ?

Therefore checking than we have available UTXOs in mempool (or in-package) to spend could be done before to call to `AncestorPackage` ctor and `visit` to obtain the `ancestor_subset`, sounds it could be a small perf again assuming `AcceptPackage` is re-worked a bit.

Reply:
> Therefore checking than we have available UTXOs in mempool (or in-package) to spend could be done before to call to AncestorPackage ctor

I don't think this is true. We should sort the transactions before we look up UTXOs.

Reply:
Fair, I still think it’s slightly less DoSy though depends benchmark and UTXO layout on disk of course.